test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,461 @@
+		t.Logf("will use existing keycloak deployment at URL: %s", keycloakURL)
+		kcClient = tc.setupKeycloakClient(testCtx, keycloakURL)
+
+	} else {
+		t.Logf("no existing keycloak deployment found; will create new")
+		var cleanups []func()
+		kcClient, cleanups = tc.setupExternalOIDCWithKeycloak(testCtx)
+		defer test.IDPCleanupWrapper(func() {
+			for _, c := range cleanups {
+				c()
+			}
+		})()
+		t.Logf("keycloak Admin URL: %s", kcClient.AdminURL())
+	}
+
+	kcAdminClient, err := kcClient.GetClientByClientID(oidcClientId)
+	require.NoError(t, err)
+	require.NotEmpty(t, kcAdminClient)
+
+	group := names.SimpleNameGenerator.GenerateName("e2e-keycloak-group-")
+	err = kcClient.CreateGroup(group)
+	require.NoError(t, err)
+
+	user := names.SimpleNameGenerator.GenerateName("e2e-keycloak-user-")
+	email := fmt.Sprintf("%s@test.dev", user)
+	password := "password"
+	err = kcClient.CreateUser(user, email, password, []string{group})
+	require.NoError(t, err)
+
+	httpClient := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}}
+
+	formData := url.Values{
+		"grant_type": []string{"password"},
+		"client_id":  []string{oidcClientId},
+		"scope":      []string{"openid"},
+		"username":   []string{user},
+		"password":   []string{password},
+	}
+
+	resp, err := httpClient.PostForm(kcClient.TokenURL(), formData)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	data, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.NotEmpty(t, data)
+
+	var authResponse oidcAuthResponse
+	err = json.Unmarshal(data, &authResponse)
+	require.NoError(t, err)
+	require.NotEmpty(t, authResponse.AccessToken)
+	require.NotEmpty(t, authResponse.IdToken)
+
+	// ==========================================
+	// Test authentication via the kube-apiserver
+	// ==========================================
+
+	// use the OIDC id_token to do a self subject review
+	kasURL := fmt.Sprintf("%s/apis/authentication.k8s.io/v1/selfsubjectreviews", tc.kubeConfig.Host)
+	ssrReq := `{"kind":"SelfSubjectReview","apiVersion":"authentication.k8s.io/v1"}`
+	req, err := http.NewRequest(http.MethodPost, kasURL, strings.NewReader(ssrReq))
+	require.NoError(t, err)
+
+	// kubernetes uses the id_token to identify the user (instead of the access_token)
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", authResponse.IdToken))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err = httpClient.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusCreated, resp.StatusCode)
+
+	body, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	var ssr authenticationv1.SelfSubjectReview
+	err = json.Unmarshal(body, &ssr)
+	require.NoError(t, err)
+	require.Equal(t, email, ssr.Status.UserInfo.Username)
+	require.Contains(t, ssr.Status.UserInfo.Groups, "system:authenticated")
```

Comment:
Need it also check the keycloak group above?

Reply:
As far as I understand, that group does not appear in the id/access token that Keycloak returns because I haven't created any group mapper in Keycloak to include that claim into the tokens. I'll see how this can be done and will try to incorporate it.

This means that currently, `system:authenticated` is the only group returned by the SSR -- this should be enough for now for the current purposes of this test.

Reply:
Could this assert on the username instead? I imagine we don't want this test to pass if the request user is authenticated by some means other than the bearer token.